spmd/compiler/api.py

```diff
@@ -171,11 +173,34 @@ def dummy_add(grad: torch.Tensor, zero: torch.Tensor) -> torch.Tensor:
         call_functions[0], node_to_obj
     )
 
+    # rewrite graph to remove redundant clone
+    traced_dispatch = _remove_clone_tensor(traced_dispatch)
+
     traced_dispatch.graph.lint()
 
     return traced_dispatch
 
 
+def _remove_clone_tensor(subgm: fx.GraphModule) -> fx.GraphModule:
+    """rewrite dummy add graph to remove clone of grad tensor"""
+    nodemap = create_graph_node_map(subgm)
+
+    clone_node = nodemap.get("clone")
+    comm_node = nodemap.get("allreduce__default")
+
+    assert comm_node is not None, "expected all_reduce comm node not present."
+    assert clone_node is not None, "expected clone node not present."
+
+    grad_tensor_node = clone_node.args[0]
+    comm_node.update_arg(0, [grad_tensor_node])
```

Comment:
this looks very specific. Do we need to verify the following few conditions?

1. `comm_node` indeed takes `clone_node` as argument
2. there is no other places using `clone_node`.

Reply:
Good points - I've added a check and two asserts to ensure the items above (comm_node takes clone_node as arg) and no other places are using clone node:
<img width="859" alt="comm_node_user_check" src="https://user-images.githubusercontent.com/46302957/204430666-3fdc9947-b308-4b9c-a249-86c6daf2c359.png">